src/Engine.sol

```diff
@@ -401,24 +435,32 @@ contract Engine is IEngine, Multicallable, EIP712, EIP7412, ERC2771Context {
     //////////////////////////////////////////////////////////////*/
 
     /// @inheritdoc IEngine
-    function execute(ConditionalOrder calldata _co, bytes calldata _signature)
+    function execute(
+        ConditionalOrder calldata _co,
+        bytes calldata _signature,
+        uint256 _fee
+    )
         external
         override
-        returns (
-            IPerpsMarketProxy.Data memory retOrder,
-            uint256 fees,
-            uint256 conditionalOrderFee
-        )
+        returns (IPerpsMarketProxy.Data memory retOrder, uint256 fees)
     {
-        /// @dev check: (1) nonce has not been executed before
-        /// @dev check: (2) signer is authorized to interact with the account
-        /// @dev check: (3) signature for the order was signed by the signer
-        /// @dev check: (4) conditions are met || trusted executor is msg sender
-        if (!canExecute(_co, _signature)) revert CannotExecuteOrder();
+        /// @dev check: (1) fee does not exceed the max fee set by the conditional order
+        /// @dev check: (2) fee does not exceed balance credited to the account
+        /// @dev check: (3) nonce has not been executed before
+        /// @dev check: (4) signer is authorized to interact with the account
+        /// @dev check: (5) signature for the order was signed by the signer
+        /// @dev check: (6) conditions are met || trusted executor is msg sender
+        if (!canExecute(_co, _signature, _fee)) revert CannotExecuteOrder();
 
         /// @dev spend the nonce associated with the order; this prevents replay
         _useUnorderedNonce(_co.orderDetails.accountId, _co.nonce);
 
+        /// @dev impose a fee for executing the conditional order
+        /// @dev the fee is denoted in ETH and is paid to the caller (conditional order executor)
+        /// @dev the fee does not exceed the max fee set by the conditional order and
+        /// this is enforced by the `canExecute` function
+        _withdrawEth(payable(_msgSender()), _co.orderDetails.accountId, _fee);
```

Comment:
the assumption here is that `useOrderedNonce` will spend the nonce and canExecute would fail if there was re-entrancy right?

Reply:
also i don't know if this is intended but you could technically re-enter with a new _co calldata and signature, but not sure if there's a vuln there because you'd still have to execute those orders

Reply:
> the assumption here is that `useOrderedNonce` will spend the nonce and canExecute would fail if there was re-entrancy right?

That is correct. This follows the checks before effects principle 

> also i don't know if this is intended but you could technically re-enter with a new _co calldata and signature, but not sure if there's a vuln there because you'd still have to execute those orders

The ability to re-enter with a new `co`, `signature`, and `fee` is understood. `canExecute` is the point of defense for any malicious behavior.